convert.go

```diff
@@ -433,6 +434,142 @@ func GetAssignToDstType(dst interface{}) (interface{}, bool) {
+	fieldIter, err := newFieldIterator(src)
+	if err != nil {
+		return err
+	}
+
+	if len(dst) != fieldIter.fieldCount {
+		return errors.Errorf("can't scan row value, number of fields don't match: row fields count=%d desired fields count=%d", fieldIter.fieldCount, len(dst))
+	}
+
+	_, fieldBytes, eof, err := fieldIter.next()
+	for i := 0; !eof; i++ {
+		if err != nil {
+			return err
+		}
+
+		binaryDecoder, ok := dst[i].(BinaryDecoder)
+		if !ok {
+			return errors.Errorf("record field doesn't implement binary decoding: %s", reflect.TypeOf(dst[i]).Name())
+		}
+
+		if err = binaryDecoder.DecodeBinary(ci, fieldBytes); err != nil {
+			return err
+		}
+
+		_, fieldBytes, eof, err = fieldIter.next()
+	}
+
+	return nil
+}
+
+// EncodeRow builds a binary representation of row values (row(), composite types)
+func EncodeRow(ci *ConnInfo, buf []byte, fields ...Value) (newBuf []byte, err error) {
+	fieldBytes := make([]byte, 0, 128)
+
+	newBuf = pgio.AppendUint32(buf, uint32(len(fields)))
+	for _, f := range fields {
+		dt, ok := ci.DataTypeForValue(f)
+		if !ok {
+			return nil, errors.Errorf("Unknown OID for %s", f)
+		}
+		newBuf = pgio.AppendUint32(newBuf, dt.OID)
+
+		if f.Get() != nil {
+			binaryEncoder, ok := f.(BinaryEncoder)
+			if !ok {
+				return nil, errors.Errorf("record field doesn't implement binary encoding: %s", reflect.TypeOf(f).Name())
+			}
+			fieldBytes, err = binaryEncoder.EncodeBinary(ci, fieldBytes[:0])
+			if err != nil {
+				return nil, err
+			}
+
+			newBuf = pgio.AppendUint32(newBuf, uint32(len(fieldBytes)))
+			newBuf = append(newBuf, fieldBytes...)
+		} else {
+			newBuf = pgio.AppendInt32(newBuf, int32(-1))
+		}
+
+	}
+	return
+}
+
+// ROW allows deconstructing row values (records and composite types) into
+// fields directly without creating your own type and implementing decoder interfaces
+func ROW(isNull *bool, fields ...interface{}) BinaryDecoderFunc {
+	return func(ci *ConnInfo, src []byte) error {
+		var record Record
+		if err := record.DecodeBinary(ci, src); err != nil {
+			return err
+		}
+
+		if record.Status == Null {
+			*isNull = true
+			return nil
+		}
+
+		if len(record.Fields) != len(fields) {
+			return errors.Errorf("can't scan row value, number of fields don't match: row fields count=%d desired fields count=%d", len(record.Fields), len(fields))
+		}
+
+		for i, f := range record.Fields {
+			if err := f.AssignTo(fields[i]); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+}
+
+func ROWEnc(isNull bool, fields ...interface{}) BinaryEncoderFunc {
+	return func(ci *ConnInfo, buf []byte) (newBuf []byte, err error) {
+		rowFields := make([]Value, 0, len(fields))
+
+		for _, f := range fields {
+			if v, ok := f.(Value); ok {
+				rowFields = append(rowFields, v)
+				continue
+			}
+
+			rt := reflect.TypeOf(f)
+			if dt, ok := ci.DataTypeForReflectType(rt); ok {
+				rowFields = append(rowFields, dt.Value)
+				continue
+			}
+
+			if rt.Kind() == reflect.Ptr { //try base type
```

Comment:
type matching can probably be done better reusing more sophisticated logic from other parts